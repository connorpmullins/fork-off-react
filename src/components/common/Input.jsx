import React from "react";
import PropTypes from "prop-types";
import { colors, typography } from "../../styles/theme";

const Input = ({
  id,
  label,
  value,
  onChange,
  error,
  required,
  multiline,
  rows,
  type,
  min,
  max,
  placeholder,
  style,
}) => {
  const inputStyles = {
    width: "100%",
    padding: "0.75rem",
    border: `1px solid ${error ? colors.error : colors.neutral.gray300}`,
    borderRadius: "0.375rem",
    fontSize: typography.fontSize.base,
    color: colors.neutral.gray900,
    backgroundColor: colors.white,
    outline: "none",
    transition: "border-color 0.2s ease",
    ...style,
  };

  const commonProps = {
    id,
    value,
    onChange,
    placeholder,
    required,
    style: inputStyles,
  };

  return (
    <div>
      {label && (
        <label
          htmlFor={id}
          style={{
            display: "block",
            marginBottom: "0.5rem",
            color: colors.neutral.gray700,
          }}
        >
          {label}
          {required && (
            <span style={{ color: colors.error, marginLeft: "0.25rem" }}>
              *
            </span>
          )}
        </label>
      )}
      {multiline ? (
        <textarea {...commonProps} rows={rows} />
      ) : (
        <input {...commonProps} type={type} min={min} max={max} />
      )}
      {error && (
        <p
          style={{
            color: colors.error,
            fontSize: typography.fontSize.sm,
            marginTop: "0.25rem",
          }}
        >
          {error}
        </p>
      )}
    </div>
  );
};

Input.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string,
  value: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  error: PropTypes.string,
  required: PropTypes.bool,
  multiline: PropTypes.bool,
  rows: PropTypes.number,
  type: PropTypes.string,
  min: PropTypes.number,
  max: PropTypes.number,
  placeholder: PropTypes.string,
  style: PropTypes.object,
};

Input.defaultProps = {
  type: "text",
  required: false,
  multiline: false,
  rows: 3,
};

export default Input;
