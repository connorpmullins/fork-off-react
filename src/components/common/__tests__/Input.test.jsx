import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import Input from "../Input";

describe("Input", () => {
  const defaultProps = {
    id: "test-input",
    value: "",
    onChange: jest.fn(),
  };

  it("renders with default props", () => {
    render(<Input {...defaultProps} />);
    const input = screen.getByRole("textbox");
    expect(input).toBeInTheDocument();
  });

  it("displays label when provided", () => {
    render(<Input {...defaultProps} label="Username" />);
    const label = screen.getByText("Username");
    expect(label).toBeInTheDocument();
  });

  it("shows required indicator when required prop is true", () => {
    render(<Input {...defaultProps} label="Username" required />);
    const requiredIndicator = screen.getByText("*");
    expect(requiredIndicator).toBeInTheDocument();
    expect(requiredIndicator).toHaveStyle({ color: "#ef4444" });
  });

  it("displays error message when error prop is provided", () => {
    const errorMessage = "This field is required";
    render(<Input {...defaultProps} error={errorMessage} />);
    const error = screen.getByText(errorMessage);
    expect(error).toBeInTheDocument();
    expect(error).toHaveStyle({ color: "#ef4444" });
  });

  it("applies error styles to input when error prop is provided", () => {
    render(<Input {...defaultProps} error="Error message" />);
    const input = screen.getByRole("textbox");
    expect(input).toHaveStyle({ borderColor: "#ef4444" });
  });

  it("handles password input type", () => {
    render(<Input {...defaultProps} type="password" />);
    const input = screen.getByDisplayValue("");
    expect(input).toHaveAttribute("type", "password");
  });

  it("applies correct aria attributes", () => {
    render(
      <Input
        {...defaultProps}
        required
        error="Error message"
        label="Username"
      />,
    );

    const input = screen.getByRole("textbox");
    expect(input).toHaveAttribute("aria-required", "true");
    expect(input).toHaveAttribute("aria-invalid", "true");
  });

  it("matches snapshot with default props", () => {
    const { container } = render(<Input {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with label and required indicator", () => {
    const { container } = render(
      <Input {...defaultProps} label="Username" required />,
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with error state", () => {
    const { container } = render(
      <Input {...defaultProps} error="This field is required" />,
    );
    expect(container).toMatchSnapshot();
  });

  it("matches snapshot with all features", () => {
    const { container } = render(
      <Input
        {...defaultProps}
        label="Username"
        required
        error="This field is required"
        type="password"
      />,
    );
    expect(container).toMatchSnapshot();
  });
});
