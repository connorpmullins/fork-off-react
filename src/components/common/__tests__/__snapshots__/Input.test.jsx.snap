// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input matches snapshot with all features 1`] = `
<div>
  <div>
    <label
      for="test-input"
      style="display: block; margin-bottom: 0.5rem; color: rgb(51, 65, 85);"
    >
      Username
      <span
        style="margin-left: 0.25rem;"
      >
        *
      </span>
    </label>
    <input
      id="test-input"
      required=""
      style="width: 100%; padding: 0.75rem; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease;"
      type="password"
      value=""
    />
    <p
      style="font-size: 0.875rem; margin-top: 0.25rem;"
    >
      This field is required
    </p>
  </div>
</div>
`;

exports[`Input matches snapshot with default props 1`] = `
<div>
  <div>
    <input
      id="test-input"
      style="width: 100%; padding: 0.75rem; border: 1px solid #cbd5e1; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease;"
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`Input matches snapshot with error state 1`] = `
<div>
  <div>
    <input
      id="test-input"
      style="width: 100%; padding: 0.75rem; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease;"
      type="text"
      value=""
    />
    <p
      style="font-size: 0.875rem; margin-top: 0.25rem;"
    >
      This field is required
    </p>
  </div>
</div>
`;

exports[`Input matches snapshot with label and required indicator 1`] = `
<div>
  <div>
    <label
      for="test-input"
      style="display: block; margin-bottom: 0.5rem; color: rgb(51, 65, 85);"
    >
      Username
      <span
        style="margin-left: 0.25rem;"
      >
        *
      </span>
    </label>
    <input
      id="test-input"
      required=""
      style="width: 100%; padding: 0.75rem; border: 1px solid #cbd5e1; border-radius: 0.375rem; font-size: 1rem; color: rgb(15, 23, 42); outline: none; transition: border-color 0.2s ease;"
      type="text"
      value=""
    />
  </div>
</div>
`;
